### PR TITLE
Adding Architecture-Level Resource Estimator Design Document

### PR DESCRIPTION
**Pull request by spencerdearman**

### CODE REVIEW

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# Architecture-Level Quantum Resource Estimator
+
+## Spencer Dearman
+
+### **Abstract**
+
+This document outlines the design and development of a architecture-level Quantum Resource Estimation (QRE) tool. Unlike existing QRE tools, which are mainly focused on quantum algorithm-level estimations, this tool aims to address the complexities involved in quantum architecture.
```
Comment (vtomole) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:7]:
```suggestion
This document outlines the design and development of an architecture-level Quantum Resource Estimation (QRE) tool. Unlike existing QRE tools, which are mainly focused on quantum algorithm-level estimations, this tool aims to address the complexities involved in quantum architecture.
```

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+### **Background and Motivation**
+
+Existing QRE tools are limited in their ability to provide insights at the architecture level, particularly when comparing Surface Codes with LRESC. Current tools just highlight the inherent advantage of LRESC in saving qubits, a conclusion that can be reached without extensive QRE analysis. This project aims to build a more advanced QRE tool that considers other factors such as Movement, Atom Reloading, Code Teleportation, Connectivity, Degree of Parallelization, and other architectural assumptions, providing more meaningful and actionable insights.
+
+ 
+### **Current Resource Estimation Limitations**
+
+_Limitations of Azure Quantum Resource Estimator and Qualtran_
+
+Both the Azure Quantum Resource Estimator (AQRE) and Qualtran are leading tools used in quantum resource estimation. However, despite their popularity, these tools are predominantly algorithm-centric. This narrow focus on algorithms presents a significant limitation when attempting to produce realistic estimations for specific and more complex quantum systems, as they largely disregard the hardware-specific characteristics that are crucial.
```
Comment (vtomole) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:19]: Talk more about how they are algorithm-centric: They out-the-box assume Surface code with Lattice surgery and MSD.

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+_Non-Functional Requirements_
+
+- **Performance:** The tool should provide results in a reasonable timeframe, even for large-scale quantum architectures.
+
+- **Scalability:** Ability to scale with increasing complexity of quantum systems.
+
+- **Usability:** User-friendly interface for easy input of parameters and interpretation of results.
+
+- **Extensibility:** Design should allow for future enhancements and the addition of new features.
```
Comment (vtomole) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:70]: Let's remove this for more specific requirements. For example, we want to be able to do a universal gateset without magic state distillation.  People should be able to switch back and forth between 3D codes, subsystem codes, concatenated codes, and piece-able fault tolerance.

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+_Hard-Coded Architectures and Inflexibility_
+
+One of the most significant challenges we encountered with these tools is their lack of flexibility. Quantum resource estimators like AQRE and BenchQ are inherently designed to optimize predefined algorithms, and they are not equipped to adjust to the dynamic needs of various quantum architectures. For example, when examining the Alice and Bob architecture, or when dealing with Low-Density Parity-Check (LDPC) codes, the limitations became apparent. The hard-coded nature of these tools meant that any deviation from the predefined architecture resulted in inefficiencies and suboptimal results. The lack of support for critical features such as movement and flip-chip architecture further highlighted the need for a more adaptable tool.
+
+ 
+### **Qubit Architecture**
```
Comment (vtomole) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:32]: Add tapped ions section

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+_Hard-Coded Architectures and Inflexibility_
+
+One of the most significant challenges we encountered with these tools is their lack of flexibility. Quantum resource estimators like AQRE and BenchQ are inherently designed to optimize predefined algorithms, and they are not equipped to adjust to the dynamic needs of various quantum architectures. For example, when examining the [Alice and Bob](https://github.com/Alice-Bob-SW/qsharp-alice-bob-resource-estimator) architecture, or when dealing with Low-Density Parity-Check (LDPC) codes, the limitations became apparent. The hard-coded nature of these tools meant that any deviation from the predefined architecture resulted in inefficiencies and suboptimal results. The lack of support for critical features such as movement and flip-chip architecture further highlighted the need for a more adaptable tool.
```
Comment (vtomole) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:31]:
```suggestion
One of the most significant challenges we encountered with these tools is their lack of flexibility. Quantum resource estimators like AQRE and BenchQ are inherently designed to optimize predefined algorithms, and they are not equipped to adjust to the dynamic needs of various quantum architectures. For example, when examining dealing with Low-Density Parity-Check (LDPC) code architectures, the limitations became apparent. The hard-coded nature of these tools meant that any deviation from the predefined architecture resulted in inefficiencies and suboptimal results. The lack of support for critical features such as movement and flip-chip architecture further highlighted the need for a more adaptable tool.
```

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+ 
+### **Current Resource Estimation Limitations**
+
+_Limitations of Azure Quantum Resource Estimator and Qualtran_
+
+Both the Azure Quantum Resource Estimator (AQRE) and Qualtran are prominent tools in quantum resource estimation. Despite their widespread use, these tools are primarily algorithm-centric, with limited flexibility. By default, they assume a surface quantum error correction code with lattice surgery and magic state distillation, offering no customization of these parameters. In our research, we identified alternative QEC configurations that diverge from MSD and lattice surgery; however, testing these configurations was not feasible due to the tools’ rigid, hard-coded options.
```
Comment (vtomole) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:19]: We want to confirm and claim that all resource estimators (BenchQ and pyLIQTR)  suffer from this problem

**Comment by vtomole** [comment]
A thesis we want to put forward is that architecture-level resource estimations are scattered everywhere and what we are putting forward is one tool that puts all of them together. Before Stim, everybody had to build their own Stim graph from scratch. A few resource estimation papers I could find all roll their own unique architecture-level estimator:

- https://scirate.com/arxiv/1905.09749
- https://scirate.com/arxiv/2012.04108
- https://scirate.com/arxiv/2401.09541
- https://scirate.com/arxiv/2306.08585

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+The choice of qubit architecture—synthetic versus neutral atom—has significant implications for QRE. Superconducting qubits often require careful consideration of connectivity due to their fixed physical layout, while neutral atoms offer more flexibility in qubit arrangement but face challenges in operation fidelity. A comprehensive QRE tool must account for these differences in architecture, adjusting resource estimations based on the type of qubit used, its coherence properties, and the specific operational constraints associated with each platform.
+
+### **Requirements**
+
+_Functional Requirements_
+
+- **Movement Analysis:** The tool must evaluate the impact of qubit movement within the architecture.
+
+- **Atom Reloading:** Consideration of the frequency and impact of atom reloading within quantum operations.
```
Comment (richrines1) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:64]: this is probably relevant to ions too - maybe just "reloading"?

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+- **Code Teleportation:** Ability to model and assess the implications of code teleportation on qubit resources.
+
+- **Connectivity:** Analysis of qubit connectivity and its effect on the overall architecture.
+
+- **Degree of Parallelization:** Estimation of the impact of parallel operations on qubit usage and efficiency.
+
+- **Custom Assumptions:** Flexibility to incorporate various architectural assumptions and parameters.
```
Comment (richrines1) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:72]: nit: imo this should be at the end of the list (or just above "seamless integration")

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+- **Universal Gateset Configuration:** The tool must support the ability to implement a universal gateset without relying on magic state distillation, allowing for more versatile quantum operations.
+
+- **Flexible QEC Code Switching:** Users should be able to switch between different quantum error correction codes, such as 3D codes, subsystem codes, concatenated codes, and piece-able fault tolerance, providing a broader range of options for optimizing quantum systems.
```
Comment (richrines1) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:76]: these two points need to be clarified/disentangled a bit imo. "code switching" refers to something specific, which doesn't encompass everything in that bullet - e.g. "piece-able fault tolerance" is a way to implement a complete gateset (and so seems like it belongs in the previous point as an alternative to magic state distillation?)

would something like this be close to what you're going for?
1. it must consider a variety of techniques for implementing a universal gateset, including:
    * magic state distillation
    * code switching
    * piece-able fault tolerance
    * lattice surgery
    * ...
2. it must support arbitrary (or at least a wide variety of) QEC codes, ...

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+_Detailed Architecture_
+
+- **Movement Analysis Subsystem:** Implements algorithms to estimate the impact of qubit movement.
+
+- **Atom Reloading Subsystem:** Models the atom reloading process and its effect on qubit availability.
+
+- **Code Teleportation Subsystem:** Evaluates teleportation strategies and their resource implications.
+
+- **Connectivity Analysis Subsystem:** Analyzes qubit connectivity patterns and their impact on the architecture.
```
Comment (richrines1) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:100]: would it make sense to make movement and code teleportation submodules of this subsystem?

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+- **Input Module:** Accepts user-defined parameters for the quantum architecture, including movement constraints, reloading rates, and connectivity graphs.
+
+- **Analysis Engine:** Core computational module that performs the QRE calculations, considering all the defined parameters and assumptions.
+
+- **Output Module:** Generates reports and visualizations, providing insights into qubit resource usage and architectural trade-offs.
+
+- **User Interface:** A GUI that allows users to interact with the tool, input data, run simulations, and view results.
+
+_Detailed Architecture_
```
Comment (richrines1) [inline supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:92]: it feels like there's something missing here - none of these subsystems actually deal with gateset implementation, or the code itself

**Review by richrines1 — COMMENTED** [review]
nice! added a few (non-binding) comments but mostly this looks great

File: supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md (hunk 8 above)
Comment (spencerdearman) [reply supermarq-benchmarks/examples/qre-challenge/Lessons-Learned.md:72]: Good point, I agree. I moved it to the end

**Review by vtomole — APPROVED** [review]
LGTM

**Pull request merged**